apps/astarte_realm_management_api/lib/astarte_realm_management_api/triggers/trigger.ex

```diff
@@ -20,21 +20,73 @@ defmodule Astarte.RealmManagement.API.Triggers.Trigger do
   use Ecto.Schema
   import Ecto.Changeset
   alias Astarte.Core.Triggers.SimpleTriggerConfig
-  alias Astarte.RealmManagement.API.Triggers.Trigger
+  alias Astarte.RealmManagement.API.Triggers.{AMQPAction, HttpAction, Trigger}
 
   @derive {Phoenix.Param, key: :name}
   @primary_key false
   embedded_schema do
     field :name, :string
-    field :action, :map
+    field :action, :any, virtual: true
+    embeds_one :amqp_action, AMQPAction
+    embeds_one :http_action, HttpAction
     embeds_many :simple_triggers, SimpleTriggerConfig
   end
 
   @doc false
-  def changeset(%Trigger{} = trigger, attrs) do
+  def changeset(%Trigger{} = trigger, attrs, opts) do
+    attrs = move_action(attrs)
+
     trigger
-    |> cast(attrs, [:name, :action])
-    |> validate_required([:name, :action])
+    |> cast(attrs, [:name])
+    |> validate_required([:name])
+    |> cast_embed(:amqp_action, required: false, with: {AMQPAction, :changeset, [opts]})
```

Comment:
Is the extra `[]` around `opts` needed? Doesn't result in a list of lists since `opts` is already a list?

Reply:
yes, we want that: we want to pass the opts list as a list.